### DISCUSSION TITLE
how to count $PARAMS?

### DISCUSSION THREAD

**Discussion by JeffMelton** [lmorg/murex#902] [Q&A] [answered]
Super confused here. I'm writing a function to load with `.murex_profile`, and I want it to execute differently based on whether an argument is passed into it. Ultimately, if it does not receive one and only one argument, it should prompt the user (me) to make a selection from `fzf`. Either the argument from stdin or the selection from `fzf` should then be assigned to a variable for the rest of the function to use. I've tried about a million things, and I've run through the documentation until I'm cross-eyed. Here's a few super minimal examples of what I'm trying to do:

```
❯ function test {
2 » $PARAMS -> count -> set arg_count
3 » out $arg_count 
4 » $arg_count == 1
5 » out
6 » exitnum
7 » }

❯ test foo
1
false
0
❯ test foo bar
2
false
0
```

I think I understand that `$PARAMS` is `json`? The docs seem to suggest that `count` should do what I want:

>Counts the number of items in a structure, be that a list, map or other object type.

And it looks like that's what it's doing, but then the comparison isn't doing what I expect it to?

```
❯ ja [bar] -> set jarray
❯ $jarray -> count
1

❯ $jarray -> count -> set jarray_count
❯ $jarray_count == 1
false
```

Thinking I might be running into some kind of data type issue, I tried casting explicitly to JSON, and I _think_ this is how to do that?

```
❯ function test {
2 » tout json $PARAMS
3 » }
❯ test foo
[
    "foo"
]
❯ function test {
2 » tout json $PARAMS -> count
3 » }
❯ test foo
1
❯ function test {
2 » tout json $PARAMS -> count -> set arg_count
3 » $arg_count == 1
4 » }
❯ test foo
false
```

What am I missing? Waving the white flag.

**Comment by lmorg** [comment]
That's a bug:
![image](https://github.com/user-attachments/assets/cbd7c619-4d1d-4d6c-81e7-3e9a278f0198)

Weird it's gone this long without being detected in either the test suites nor every day usage.

It shouldn't be too hard to fix though. I have a feeling I know the cause.

Do you mind if I convert this discussion to an "issue"?

**Reply by JeffMelton** [reply]
Go for it. Glad to know I'm not losing my mind. 😆

**Reply by JeffMelton** [reply]
In the meantime, do you have any suggestions for workarounds? more idiomatic ways of approaching the problem?

**Reply by lmorg** [reply]
Wrapping the integer in quotation marks will work. You shouldn't need to do this though
![image](https://github.com/user-attachments/assets/73199e9f-02e1-48ab-a7e1-2c1e8b66e22b)

**Comment by lmorg** [comment, marked as answer]
@JeffMelton 

I've found the issue. `count` is returning an `int` while numeric constants in expressions are `num` (floating point numbers).

eg
```
murex » %[1..3] -> count -> set num arg_count
murex » $arg_count == 3
true
murex » %[1..3] -> count -> set arg_count
murex » $arg_count == 3
false
```

I still consider this a bug because by default Murex is relaxed when comparing numeric types (hence why comparing an `int` to a `str` works) so it should just compare the value of those two numeric types. But it does explain why this bug hasn't been picked up before: `int` is rarely used since `num` is the default numeric data type.

I'll get a fix for this in the next release.

**Reply by lmorg** [reply]
Fix in `develop` branch. This was literally just a one line change too. Pity not all bugs are this easy 😆 
https://github.com/lmorg/murex/pull/900/commits/ab1f24fe8d9b79d71a5394b53f3d0609ae281ca3

**Reply by JeffMelton** [reply]
Nice. Thanks for being so responsive. 😊
